hypothesis-python/src/hypothesis/internal/cache.py

```diff
@@ -180,7 +225,16 @@ def __out_of_order(self, i, j):
         i must be the parent of j.
         """
         assert i == (j - 1) // 2
-        return self.data[j].score < self.data[i].score
+        di = self.data[i]
+        dj = self.data[j]
+
+        # Always place pinned entries below unpinned entries. Beyond that,
+        # don't worry about the ordering of pinned entries.
+        if di.pins > 0:
+            return dj.pins == 0
+        if dj.pins > 0:
+            return False
+        return dj.score < di.score
```

Comment:
This logic can potentially be simplified by a suitable helper method in `Entry`:

```python
    @property
    def sort_key(self):
        if self.pins == 0:
            # Unpinned entries are sorted by score.
            return (0, self.score)
        else:
            # Pinned entries sort after unpinned ones. Beyond that, we don't
            # worry about their relative order.
            return (1, 0)
```

That way we can just write `self.data[j].sort_key < self.data[i].sort_key`.

(This also applies to the assertion in `__swap`.)